packages/drivers/driver-base/src/documentDeltaConnection.ts

```diff
@@ -305,16 +315,41 @@ export class DocumentDeltaConnection
         this.submitManager.add("submitSignal", [message]);
     }
 
+    /**
+     * Disconnect from the websocket, and permanently disable this DocumentDeltaConnection.
+     */
+    public close() {
```

Comment:
In most of our objects the `close()` method permanently disables/disposes the object (the comment here implies that also).  However, it doesn't look like we enforce any permanent disable here and I see `canRetry` is `true` below.  Maybe rename to keep the semantic of "close" consistent in the repo.

Reply:
I'm not sure what we can improve here. Could you please clarify?
As a general statement, subclassing can break any invariant for any class, I do not think there is a cure against that :)

Reply:
I think I don't really understand why `disconnectCore` was split out from `disconnect`, I only see `disconnectCore` getting overridden with this change, is there a scenario for overriding `disconnect` rather than `disconnectCore` with this change?

If not, then I think merging `disconnect` into `close` but keeping `disconnectCore` split out brings it back in line with our other `close` methods.  Also maybe `canRetry` should be false?

Reply:
The only difference I split disconnect & close is to be more assertive about need to provide reason for disconnect. Same can be done with single method and default args, but splitting seems cleaner - close() is for external users, disconnect() for internal and they have to provide reason, it can't be some default reason. disconnectCore is the one that is supposed to be overwritten only.
I can merge them back together if you feel like it's more readable / less likely for people to make mistakes.

Reply:
Agree the split is good, just suggesting 2 methods rather than 3:

```typescript
public close() // don't override this, sets the closed flag
protected disconnect(socketProtocolError: boolean, reason: DriverError) // good to override, by default disconnects the socket
```

Reply:
Are you sure? :)
I really do not want to duplicate that code across 2 implementations:
```
    protected disconnect(socketProtocolError: boolean, reason: DriverError) {
        // We set the closed flag as a part of the contract for overriding the disconnect method. This is used by
        // DocumentDeltaConnection to determine if emitting on the socket is allowed, which is important since
        // OdspDocumentDeltaConnection reuses the socket rather than truly disconnecting it.  Note that below we may
        // still send disconnect_document which is allowed; this is only intended to prevent normal messages from
        // being emitted.
        if (this.closed) {
            return;
        }
        this.closed = true;

        this.removeTrackedListeners(false);
```

Reply:
Or are you suggesting moving it into close?

Reply:
this.close handling is probably fine being moved to close, this.removeTrackedListeners(false) call would need to be duplicated that way

Reply:
The latter, move it into close (so then close is the one setting the closed flag, ensuring the object is permanently disabled even if the protected method is overridden)

Reply:
Take a look at latest iteration - I reorganized how we react to various errors by introducing fail() callback.
The weak spot - if someone calls disconnect() directly, they are not clearing callbacks...